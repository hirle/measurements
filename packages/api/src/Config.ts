
export interface LogsConfig {
    dir: string
    retention: number
    level?: string
  }

export interface DatabaseConfig {
  type: string,
  config: unknown
} 

export interface RecorderConfig {
  id: string,
  "measurement-id": string,
  mode: string,
  config?: unknown
} 

export interface SensorConfig {
  id: string,
  type: string,
  config: unknown
}

export interface MeasurementSupplierConfig {
  id: string,
  "sensor-id": string,
  "sensor-key": string
}

export default interface Config {
  "http-port": number,
  database: DatabaseConfig,
  sensors: SensorConfig[],
  measurements: MeasurementSupplierConfig[],
  recorders: RecorderConfig[],
  logs?: LogsConfig
}

