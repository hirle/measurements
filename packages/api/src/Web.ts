import express, { RequestHandler } from 'express';
import { json as BodyParserJson }  from 'body-parser';
import { Server as HttpServer } from 'http';
import { Server as socketIOServer} from 'socket.io';
import { getLogger as Log4jsGetLogger, connectLogger as Log4jsConnectLogger } from 'log4js';

export default class Web {

    private httpPort: number;
    private app: express.Application;
    private httpServer: HttpServer;
    private io: socketIOServer;
    private static webLogger = Log4jsGetLogger('web');

    constructor(httpPort: number) {
        this.app = express()
        this.httpPort = httpPort;
        this.httpServer = new HttpServer(this.app)
        this.io = new socketIOServer(this.httpServer);
    }

    startOn( ) {

        this.app.use(Log4jsConnectLogger(Web.webLogger, { level: "info" }));

        this.app.use(BodyParserJson());  

        // TODO serve static files

        this.httpServer.listen(this.httpPort, () => {
            Web.webLogger.info(`Listening on ${this.httpPort}`);
        });
    }

    recordGetRoute(path: string, requestHandler: RequestHandler ): void {
        this.app.get(path, requestHandler);
        Web.webLogger.trace(`Record get on ${path}`);
    }

    recordPostRoute(path: string, requestHandler: RequestHandler ): void {
        this.app.post(path, requestHandler);
        Web.webLogger.trace(`Record get on ${path}`);
    }
}