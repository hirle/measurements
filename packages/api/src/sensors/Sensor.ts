import ObjectWithId, { ObjectWithIdCollection } from "../patterns/ObjectWithIdCollection";
import { Unit } from "../Unit";
import { isValueType, ValueType } from "../ValueType";


export abstract class Sensor implements ObjectWithId{
  public readonly id: string;

  protected constructor( id:string ) {
    this.id = id;
  }

  public abstract fetchValue(): Promise<SensorValues>;
  public abstract getValuesKeys(): Readonly<Set<string>>;
  public abstract getKeyValueUnit( key:string): Readonly<Unit>;
}


export interface SensorValues {
  values: Map<string, SensorValue>
} 

export interface SensorValue {
  at: Date,
  value: ValueType
}

export function narrowSensorValue ( maybeSensorValue: unknown ): SensorValue {
  if (isAltmostSensorValue ( maybeSensorValue )
  &&  isValueType(maybeSensorValue.value)
  &&  mayBecomeADate( maybeSensorValue.at ) ) {
    return { at: new Date((maybeSensorValue).at), value: maybeSensorValue.value};
  } else {
    throw new Error('unexpected data')
  } 
}

type AltmostSensorValue = { at: unknown, value: unknown };
function isAltmostSensorValue ( maybeSensorValue: unknown ): maybeSensorValue is AltmostSensorValue {
  return typeof maybeSensorValue === 'object'
  && maybeSensorValue !== null
  && 'at' in maybeSensorValue
  && 'value' in maybeSensorValue;
}

function mayBecomeADate( input: unknown ) : input is string|number|Date {
  return input && ( ['string', 'number'].includes(typeof input) ||Â input instanceof Date );
}


export class SensorCollection extends ObjectWithIdCollection<Sensor>{}