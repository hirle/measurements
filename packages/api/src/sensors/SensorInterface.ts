export interface SensorInterface {
  fetchValue(): Promise<SensorValues>;
  getValuesKeys(): Readonly<string[]>;
}

export type SensorValueType = boolean|number|string;

export interface SensorValues {
  timestamp: Date,
  values: Map<string, SensorValueType>
} 