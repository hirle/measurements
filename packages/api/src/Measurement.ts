import ObjectWithId, { ObjectWithIdCollection } from "./patterns/ObjectWithID";
import { Sensor, SensorValues } from "./Sensor";
import Supplier from "./patterns/Supplier";
import { Unit } from "./Unit";
import { ValueType } from "./ValueType";

export interface Measurement {
  timestamp: Date,
  value: ValueType,
  unit?: Unit;
}

export class MeasurementSupplier implements Supplier<Promise<Measurement>>, ObjectWithId {
  
  public readonly id;
  public readonly sensor: Sensor;
  public readonly key: string;

  public constructor(id:string, sensor: Sensor, key: string) {
    this.id = id;
    this.sensor = sensor;
    this.key = key;
  }

  public get(): Promise<Measurement> {
    return this.sensor.fetchValue()
      .then( ( sensorValues: SensorValues )  => {
        return sensorValues.values.get(this.key);
      });
  }
}

export class MeasurementSupplierCollection extends ObjectWithIdCollection<MeasurementSupplier> {
}