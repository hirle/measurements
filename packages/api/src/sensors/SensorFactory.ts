import { Sensor } from '../Sensor';
import { SensorConfig } from '../Config';
import { TCW122Sensor } from './TCW122Sensor';

export class SensorFactory  {

  public static create( config: SensorConfig ) : Sensor {
    switch( config.type ) {
      case 'tcw122': return TCW122Sensor.create(config.id, config.config);
      case '1wire': throw new Error('Not implemented yet');
      default: throw new Error('Unknown sensor type');
    } 
  }
}