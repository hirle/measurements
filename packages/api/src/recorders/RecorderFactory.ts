import { RecorderConfig } from "../Config";
import { MeasurementSupplier } from "../Measurement";
import MeasurementDatabase from "../MeasurementDatabase";
import ManualRecorder from "./ManualRecorder";
import PeriodicRecorderFactory from "./PeriodicRecorderFactory";
import Recorder from "./Recorder";

export default class RecorderFactory  {
  public static create(
    recordingConfig: RecorderConfig,
    measurementSupplier: MeasurementSupplier,
    measurementDb: MeasurementDatabase
  ) : Recorder {
    switch( recordingConfig.mode ) {
      case 'manual': return new ManualRecorder(recordingConfig.id, measurementSupplier, measurementDb)
      case 'periodic': return PeriodicRecorderFactory.create(
        recordingConfig.id,
        measurementSupplier,
        measurementDb,
        recordingConfig.config);
      default: throw new Error(`unknown recorder mode ${recordingConfig.mode}`)
    }
  }
}