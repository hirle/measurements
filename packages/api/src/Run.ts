import Config, { SensorConfig } from './Config';
import DefaultConfig from './default.config.json';
import fs from 'fs';
import Logger from "./Logger";
import Web from "./Web";
import GetVersion from './GetVersion';
import SupplierHandler from './SupplierHandler';
import { ApiVersionInterface } from '@measures/apiinterface';
import { SensorInterface } from './sensors/SensorInterface';
import { SensorFactory } from './sensors/SensorFactory';

export function run(argv: string[]): number {
    
    const config = processArgv(argv);

    const logger: Logger = Logger.create(config.logs);

    const _sensors: SensorInterface[] = setUpSensors(config.sensors);

    // TO DO give sensors to schedulers

    const web = new Web(config['http-port']);
    
    web.startOn()
    
    setupApiRoutes(web);

    logger.info('Ready!');

    return 0;
}

function setUpSensors(sensors: SensorConfig[]): SensorInterface[] {
  return sensors.map( sensorConfig => SensorFactory.create(sensorConfig));
}

function setupApiRoutes( web: Web ) {

  // TODO set to process.env.npm_package_version
  const getVersion = new GetVersion('0.0.1');
  web.recordGetRoute('/api/version', SupplierHandler.create<ApiVersionInterface>(getVersion));
}

function processArgv(argv: string[]): Config {
  switch( argv.length ) 
  {
    case 2: return DefaultConfig; 
    case 3: throw new Error('Missing argument: ./path/to/config.json');
    case 4: if( argv[2] === '--config' ) {
        return JSON.parse(fs.readFileSync(argv[3], 'utf8'))
      }  else {
        throw new Error('Bad argument');
      } 
    default:  
      throw new Error('Bad argument');
  }
}
