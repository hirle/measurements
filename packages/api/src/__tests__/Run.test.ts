import {run} from '../Run';
import Web from '../Web';
import { SensorFactory } from '../sensors/SensorFactory';
import RecorderFactory from '../recorders/RecorderFactory';
import * as MeasurementModule from '../Measurement';
import * as Log4js from 'log4js';
import { LogsConfig } from '../Config';
jest.mock('../Web');
jest.mock('../MeasurementsDatabase');
jest.mock('log4js');

describe('Run', ()=>{
  it('should run with default ', ()=>{;

    const mockedLog4jsconfiguration = mockFunction(Log4js.configure);
    const mockedloggerInfo = jest.fn<void,[string]>();
    const mockedLog4jsgetLogger = mockFunction(Log4js.getLogger)
      .mockImplementation( () => ({info: mockedloggerInfo} as unknown as Log4js.Logger));

    const spyOnSensorFactory = jest.spyOn(SensorFactory, 'create');
    const spyOnMeasurementSupplierConstructor = jest.spyOn(MeasurementModule.MeasurementSupplier, 'create' );
    const spyOnRecorderFactory = jest.spyOn(RecorderFactory, 'create');

    // run with default
    run(['node', 'index.js']);
    
    expect(mockedLog4jsconfiguration.mock.calls).toMatchSnapshot();
    expect(mockedLog4jsgetLogger.mock.calls).toMatchSnapshot();
    expect(mockedloggerInfo.mock.calls).toMatchSnapshot();

    expect(spyOnSensorFactory).toBeCalledTimes(1);
    expect(spyOnRecorderFactory).toBeCalledTimes(1);
    expect(spyOnMeasurementSupplierConstructor).toBeCalled();
    expect(Web).toBeCalledTimes(1);
  });
});

// https://instil.co/blog/typescript-testing-tips-mocking-functions-with-jest/
function mockFunction<T extends (...args: any[]) => any>(fn: T): jest.MockedFunction<T> {
  return fn as jest.MockedFunction<T>;
}