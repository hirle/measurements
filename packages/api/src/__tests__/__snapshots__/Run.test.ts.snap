// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Run should log the creation of sensor, measurement supplier and recorder, should log the creation of routes  1`] = `
Array [
  Array [
    Object {
      "appenders": Object {
        "appDateRollingFile": Object {
          "filename": "logs/app.log",
          "keepFileExt": true,
          "numBackups": 7,
          "pattern": ".yyyy-MM-dd",
          "type": "dateFile",
        },
        "stdout": Object {
          "type": "stdout",
        },
        "webDateRollingFile": Object {
          "filename": "logs/web.log",
          "keepFileExt": true,
          "numBackups": 7,
          "pattern": ".yyyy-MM-dd",
          "type": "dateFile",
        },
      },
      "categories": Object {
        "app": Object {
          "appenders": Array [
            "appDateRollingFile",
          ],
          "level": "INFO",
        },
        "default": Object {
          "appenders": Array [
            "stdout",
          ],
          "level": "WARN",
        },
        "web": Object {
          "appenders": Array [
            "webDateRollingFile",
          ],
          "level": "INFO",
        },
      },
    },
  ],
]
`;

exports[`Run should log the creation of sensor, measurement supplier and recorder, should log the creation of routes  2`] = `
Array [
  Array [
    "app",
  ],
  Array [],
  Array [],
  Array [
    "app",
  ],
  Array [],
  Array [],
]
`;

exports[`Run should log the creation of sensor, measurement supplier and recorder, should log the creation of routes  3`] = `
Array [
  Array [
    "1 sensors set up",
  ],
  Array [
    "1 measurement supplier set up",
  ],
  Array [
    "1 recorders set up",
  ],
  Array [
    "Web paths set up",
  ],
]
`;

exports[`Run should log the creation of sensor, measurement supplier and recorder, should log the creation of routes  4`] = `
Array [
  Array [
    "/api/version",
    [Function],
  ],
  Array [
    "/api/measurement/teracom-demo-temperature/current",
    [Function],
  ],
  Array [
    "/api/recorder/teracom-demo-temperature-recorder/measurements/latest(?:/:count([0-9]+))?",
    [Function],
  ],
]
`;

exports[`Run should log the creation of sensor, measurement supplier and recorder, should log the creation of routes  5`] = `
Array [
  Array [
    "/api/recorder/teracom-demo-temperature-recorder/recordOneMeasurement",
    [Function],
  ],
  Array [
    "/api/recorder/teracom-demo-temperature-recorder/startRecording",
    [Function],
  ],
  Array [
    "/api/recorder/teracom-demo-temperature-recorder/stopRecording",
    [Function],
  ],
]
`;
